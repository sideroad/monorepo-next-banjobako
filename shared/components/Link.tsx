import React, { FC } from 'react';
import NextLink from 'next/link';
import qs from 'qs';

interface Props {
  path: string;
  query?: { [x in string]: any };
  children: React.ReactElement;
}

const normalize = (_uri: string, _params: { [x: string]: string }) => {
  let uri = _uri;
  const params = {};
  Object.keys(_params || {}).forEach(key => {
    if (uri.match(`\\[${key}\\]`)) {
      uri = uri.replace(
        new RegExp(`\\[${key}\\]`, 'g'),
        encodeURIComponent(_params[key])
      );
    } else {
      params[key] = _params[key];
    }
  });
  return { uri, params };
};

const Link: FC<Props> = (props: Props) => {
  const { uri: as, params } = normalize(props.path, props.query);
  const query = qs.stringify(params);

  return (
    <NextLink
      {...props}
      href={`${props.path}${query ? `?${query}` : ''}`}
      as={`${as}${query ? `?${query}` : ''}`}
    >
      {props.children}
    </NextLink>
  );
};

export default Link;
