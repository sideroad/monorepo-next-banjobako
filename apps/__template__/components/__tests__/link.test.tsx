import React from 'react';
import Link from '../link';
import renderer from 'react-test-renderer';
import { Provider, init } from '@sideroad/react-i18n';

test('Link to render', () => {
  window.__SIMPLE_I18N__ = { resource: {}, lang: 'en' };

  const i18n = init({
    assignedLanguage: 'en'
  });
  const component = renderer.create(
    <Provider value={i18n}>
      <Link href="/">
        <a>Hello World</a>
      </Link>
    </Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Link to render with as', () => {
  window.__SIMPLE_I18N__ = { resource: {}, lang: 'en' };

  const i18n = init({
    assignedLanguage: 'en'
  });
  const component = renderer.create(
    <Provider value={i18n}>
      <Link href="/users/[id]" as="/users/sideroad">
        <a>Hello World</a>
      </Link>
    </Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
