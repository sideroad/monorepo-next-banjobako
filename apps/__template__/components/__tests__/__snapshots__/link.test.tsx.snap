// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link to render 1`] = `
<a
  href="/en"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  Hello World
</a>
`;

exports[`Link to render with as 1`] = `
<a
  href="/en/users/sideroad"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  Hello World
</a>
`;
